keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -355,6 +423,193 @@ class ClusterHandler(
+   * current serverGroup.
+   *
+   * Scaling policies are treated as immutable by keel once applied. If an existing
+   * policy is modified, it will be deleted and reapplied via a single task.
+   */
+  private fun ResourceDiff<ServerGroup>.modifyScalingPolicyJob(startingRefId: Int = 0): List<Map<String, Any?>> {
+    var (refId, stages) = toDeletePolicyJob(startingRefId)
+    val newTargetPolicies = when (current) {
+      null -> desired.scaling.targetTrackingPolicies
+      else -> desired.scaling.targetTrackingPolicies
+        .subtract(current!!.scaling.targetTrackingPolicies)
+    }
+    val newStepPolicies = when (current) {
+      null -> desired.scaling.stepScalingPolicies
+      else -> desired.scaling.stepScalingPolicies
+        .subtract(current!!.scaling.stepScalingPolicies)
+    }
+
+    if (newTargetPolicies.isNotEmpty()) {
+      val (newRef, jobs) = newTargetPolicies.toCreateJob(refId, current ?: desired)
+      refId = newRef
+      stages.addAll(jobs)
+    }
+
+    if (newStepPolicies.isNotEmpty()) {
+      stages.addAll(newStepPolicies.toCreateJob(refId, current ?: desired))
+    }
+
+    return stages
+  }
+
+  private fun ResourceDiff<ServerGroup>.toDeletePolicyJob(startingRefId: Int):
+    Pair<Int, MutableList<Map<String, Any?>>> {
+    var refId = startingRefId
+    val stages: MutableList<Map<String, Any?>> = mutableListOf()
+    if (current == null) {
+      return Pair(refId, stages)
+    }
+    val current = current!!
+    val targetPoliciesToRemove = current.scaling.targetTrackingPolicies.filterNot {
+      desired.scaling.targetTrackingPolicies.contains(it)
+    }
+    val stepPoliciesToRemove = current.scaling.stepScalingPolicies.filterNot {
+      desired.scaling.stepScalingPolicies.contains(it)
+    }
+    val policyNamesToRemove = targetPoliciesToRemove.mapNotNull { it.name } +
+      stepPoliciesToRemove.mapNotNull { it.name }
+        .toSet()
+
+    stages.addAll(
+      policyNamesToRemove
+        .map {
+          refId++
+          mapOf(
+            "refId" to refId.toString(),
+            "requisiteStageRefIds" to when (refId) {
+              0, 1 -> listOf()
+              else -> listOf((refId - 1).toString())
+            },
+            "type" to "deleteScalingPolicy",
+            "policyName" to it,
+            "cloudProvider" to CLOUD_PROVIDER,
+            "credentials" to desired.location.account,
+            "moniker" to mapOf(
+              "app" to current.moniker.app,
+              "stack" to current.moniker.stack,
+              "detail" to current.moniker.detail,
+              "cluster" to current.moniker.name,
+              "sequence" to current.moniker.sequence
+            ),
+            "region" to current.location.region
+          )
+        }
+        .toMutableList()
+    )
+
+    return Pair(refId, stages)
+  }
+
+  private fun Set<TargetTrackingPolicy>.toCreateJob(startingRefId: Int, serverGroup: ServerGroup):
```

Comment:
I wonder if we should pull all these `.toXJob(...)` functions out to a separate file to make this handler easier to reason about... Thoughts? Not needed for this PR.

Reply:
+1 to that idea, but I had a Kotlin question: if you define these extension functions in another file, would they still have access to `ClusterHandler` members somehow?

Reply:
Does it have any utility outside of the one handler though? It would make sense if we re-used these functions but I don't think we would.

Reply:
Yeah, probably not, but at least for me the feeling is that some files end up with a ton of extension functions which are not exactly part of the "meat" of the class but still add to the lines of code in the file. I, for one, am still trying to get a better feel for what the "right" balance is for using extension functions versus adding them to the actual files of the target types (but for the most part the use case I see is relying on internals of the type where you define the extension, hence my question above).